### PR TITLE
RPL mesh multi-node sample using qemu

### PR DESCRIPTION
**Pull request by pedropaulomartucci**
@rveerama1 as you requested, this is my setup for RPL global repair testing.

Hope it helps to reproduce the failures.

Thank you!

### CODE REVIEW

**Comment by pfalcon** [comment]
@pedropaulomartucci : Thanks for the patch! As you can see, it now has conflicts with the latest master, and its build was failing. Please rebase and fix build issues.

**Comment by pedropaulomartucci** [comment]
@pfalcon : Done.

**Comment by pedropaulomartucci** [comment]
@pfalcon : I will take a look in the guidelines latter, probably the commit messages are wrong as you said.

**Comment by rveerama1** [comment]
@pedropaulomartucci : I am trying to utilize this PR to validate different RPL mesh scenarios on top of https://github.com/zephyrproject-rtos/zephyr/pull/980 and https://github.com/zephyrproject-rtos/zephyr/pull/5035 PRs. And hub.c is using 'glib'. We usually keep these files in net-tools (https://github.com/zephyrproject-rtos/net-tools). There are some changes needed, I will update you about my progress.

**Comment by pedropaulomartucci** [comment]
@rveerama1 : I can remove the hub application from PR and move it to net-tools repository if you think is a good call, tell me.

About the RPL mesh scenarios, keep me updated. I did some tests last week with this PR on top of #980 and #5035 and noted some misbehave when pinging between nodes, even with the routes apparently right.

I will run some more tests to try reaching the root cause.

**Comment by pedropaulomartucci** [comment]
@rveerama1 : As you suggested, I send a PR with the virtual-hub tool to the net-tools repository:
https://github.com/zephyrproject-rtos/net-tools/pull/17

The original purpose of this PR was setup RPL network scenario, which triggers a specific global-repair error. In the process, I developed the above tool, which allows setting up different network topologies for RPL using QEMU.

This way, I updated the original sample to be a guide of how to use the virtual-hub and setup multiple node RPL scenarios.

File: drivers/ieee802154/ieee802154_uart_pipe.c [hunk 1]
```diff
@@ -24,9 +24,74 @@
...
+
 /** Singleton device used in uart pipe callback */
 static struct device *upipe_dev;
 
+#if !defined(CONFIG_IEEE802154_UPIPE_PROMISCUOUS)
+
+static bool received_dest_addr_matched(u8_t *rx_buffer)
+{
+	struct upipe_context *upipe = upipe_dev->driver_data;
+	struct ieee802154_context *ctx = net_if_l2_data(upipe->iface);
```
Comment (tbursztyka) [inline drivers/ieee802154/ieee802154_uart_pipe.c]: what belongs to L2, stays in L2. So you'll have to manage this differently. 

As you are implement hw filtering, L2 will take advantage of it and will send the ext addr, short addr, and pan_id to you driver: so you'll store them and use them for matching packets. No need to grab l2's context.

File: drivers/ieee802154/Kconfig [hunk 2]
```diff
@@ -66,4 +66,48 @@ config IEEE802154_UPIPE_DRV_NAME
 	default "IEEE802154_UPIPE"
 	depends on IEEE802154_UPIPE
 
+config IEEE802154_UPIPE_PROMISCUOUS
```
Comment (tbursztyka) [inline drivers/ieee802154/Kconfig]: Change the name here. What you have been implementing is not 15.4 promiscuous mode but "hw filtering".

See include/net/ieee802154_radio.h 

With your modifications, the upipe driver could return IEEE802154_HW_FILTER as hw capabilities.

File: drivers/ieee802154/ieee802154_uart_pipe.c [hunk 3]
```diff
@@ -24,9 +24,74 @@
...
+
+	/* Check destination PAN Id */
+	if (memcmp(&rx_buffer[PAN_ID_OFFSET],
+			(u8_t *) &ctx->pan_id, PAN_ID_SIZE) != 0 &&
```
Comment (tbursztyka) [inline drivers/ieee802154/ieee802154_uart_pipe.c]: parameters are always aligned to last '(' + 1 space. 

Apply that comment everywhere

File: drivers/ieee802154/ieee802154_uart_pipe.c [hunk 4]
```diff
@@ -24,9 +24,74 @@
...
+		if (memcmp(&rx_buffer[DEST_ADDR_OFFSET],
+				BROADCAST_ADDRESS,
+				SHORT_ADDRESS_SIZE != 0)
+			&& (upipe->iface->link_addr.len != SHORT_ADDRESS_SIZE ||
+				memcmp(&rx_buffer[DEST_ADDR_OFFSET],
+						(u8_t *) &ctx->short_addr,
+						SHORT_ADDRESS_SIZE) != 0)) {
+			return false;
+		}
+	break;
```
Comment (tbursztyka) [inline drivers/ieee802154/ieee802154_uart_pipe.c]: wrong indent here, one tab more is needed. Apply that where relevant, there are other case below.

File: drivers/ieee802154/ieee802154_uart_pipe.c [hunk 5]
```diff
@@ -24,9 +24,74 @@
...
+			BROADCAST_ADDRESS, PAN_ID_SIZE) != 0) {
+		return false;
+	}
+
+	/* Check destination address */
+	switch (rx_buffer[DEST_ADDR_TYPE_OFFSET] & DEST_ADDR_TYPE_MASK) {
+	case DEST_ADDR_TYPE_SHORT:
+		/* First check if the destination is broadcast */
+		/* If not broadcast, check if lenght and address matches */
+		if (memcmp(&rx_buffer[DEST_ADDR_OFFSET],
```
Comment (tbursztyka) [inline drivers/ieee802154/ieee802154_uart_pipe.c]: if you want you could move that to a dedicated inline function

**Review by tbursztyka — CHANGES_REQUESTED** [review]
Nice addition.

some style issues, but also take a look at hw filtering flag on 153.4 radio api, that will make things nicer.

**Comment by pedropaulomartucci** [comment]
@tbursztyka : I fixed your requirements. I don't know if I did exactly what you requested about IEEE802154_HW_FILTER, tell me if any other change is required.

I had some trouble after rebasing my PR to a recent commit from master. Apparently the main problem is related with the new function: `ieee802154_enable`. Inside it, context channel is checked to not be IEEE802154_NO_CHANNEL. On the other hand, `ieee802154_init` initialized context channel value with IEEE802154_NO_CHANNEL.

For me the ideia looks ok, if we have a way to configure the channel before trying to enable the interface.

The point is, even setting up `CONFIG_NET_APP_IEEE802154_DEV_NAME="IEEE802154_UPIPE"` to trigger properly `ieee802154_settings.h` configuration function, the `ieee802154_enable` is called before it, and the channel check cause a failure when enabling the interface.

Apparently there is a tiny synchronization problem. Am I missing something?

About this PR: To validate this PR changes after rebasing I temporarily removed the IEEE802154_NO_CHANNEL check, and everything seems to work.

**Comment by tbursztyka** [comment]
@pedropaulomartucci Sorry for the delay. About the init and enable function in L2, yes it's normal. Reason being that: channel must be properly configured externally (through net_mgmt interface) before putting the device up (via net_if_up() ). We had if up/down issues with ipv6 ND/DAD before.

Your PR seems fine to me. However, last request (looks like I forgot to ask it last time), could you split it in 2 patches: one adding hw filtering to 15.4 upipe driver, and one adding your sample? Thanks.

**Comment by pedropaulomartucci** [comment]
@tbursztyka No problem, I will split my patch.

**Review by tbursztyka — APPROVED** [review]

**Review by jukkar — APPROVED** [review]
LGTM

**Pull request merged**